bridges/sloghandler/handler.go

```diff
@@ -0,0 +1,360 @@
+// Copyright The OpenTelemetry Authors
+// SPDX-License-Identifier: Apache-2.0
+
+// Package sloghandler provides a bridge between the [log/slog] and
+// OpenTelemetry logging.
+package sloghandler // import "go.opentelemetry.io/contrib/bridges/sloghandler"
+
+import (
+	"context"
+	"fmt"
+	"log/slog"
+	"slices"
+	"sync"
+
+	"go.opentelemetry.io/otel/log"
+	"go.opentelemetry.io/otel/log/noop"
+)
+
+const (
+	bridgeName = "go.opentelemetry.io/contrib/bridge/sloghandler"
+	// TODO: hook this into the release pipeline.
+	bridgeVersion = "0.0.1-alpha"
+)
+
+type config struct{}
+
+// Option configures a [Handler].
+type Option interface {
+	apply(config) config
+}
+
+// Handler is an [slog.Handler] that sends all logging records it receives to
+// OpenTelemetry.
+type Handler struct {
+	// Ensure forward compatibility by explicitly making this not comparable.
+	noCmp [0]func() //nolint: unused  // This is indeed used.
+
+	attrs  *kvBuffer
+	group  *group
+	logger log.Logger
+}
+
+// Compile-time check *Handler implements slog.Handler.
+var _ slog.Handler = (*Handler)(nil)
+
+// New returns a new [Handler] to be used as an [slog.Handler].
+func New(lp log.LoggerProvider, _ ...Option) *Handler {
+	if lp == nil {
+		// Do not panic.
+		lp = noop.NewLoggerProvider()
+	}
+	return &Handler{
+		logger: lp.Logger(
+			bridgeName,
+			log.WithInstrumentationVersion(bridgeVersion),
+		),
+	}
+}
+
+// Handle handles the passed record.
+func (h *Handler) Handle(ctx context.Context, record slog.Record) error {
+	h.logger.Emit(ctx, h.convertRecord(record))
+	return nil
+}
+
+func (h *Handler) convertRecord(r slog.Record) log.Record {
+	var record log.Record
+	record.SetTimestamp(r.Time)
+	record.SetBody(log.StringValue(r.Message))
+
+	const sevOffset = slog.Level(log.SeverityDebug) - slog.LevelDebug
+	record.SetSeverity(log.Severity(r.Level + sevOffset))
+
+	if h.attrs.Len() > 0 {
+		record.AddAttributes(h.attrs.KeyValues()...)
+	}
+
+	n := r.NumAttrs()
+	if h.group != nil {
+		if n > 0 {
+			buf, free := getKVBuffer()
+			defer free()
+			r.Attrs(buf.AddAttr)
+			record.AddAttributes(h.group.KeyValue(buf.KeyValues()...))
+		} else {
+			// A Handler should not output groups if there are no attributes.
+			g := h.group.NextNonEmpty()
+			if g != nil {
+				record.AddAttributes(g.KeyValue())
+			}
+		}
+	} else if n > 0 {
+		buf, free := getKVBuffer()
+		defer free()
+		r.Attrs(buf.AddAttr)
+		record.AddAttributes(buf.KeyValues()...)
+	}
+
+	return record
+}
+
+// Enable returns true if the Handler is enabled to log for the provided
+// context and Level. Otherwise, false is returned if it is not enabled.
+func (h *Handler) Enabled(context.Context, slog.Level) bool {
+	// TODO (MrAlias): The Logs Bridge API does not provide a way to retrieve
+	// the current minimum logging level yet.
+	// https://github.com/open-telemetry/opentelemetry-go/issues/4995
+	return true
```

Comment:
Even without https://github.com/open-telemetry/opentelemetry-go/issues/4995 I think that the bridge should accept an option to set the minimum slog.Level. Take notice that a user may want to use a single exporter and log provider while its code base can use already use multiple logging libraries that are configured on different levels (this happens in legacy systems).

I think such option would be more important and usable than https://github.com/open-telemetry/opentelemetry-go/issues/4995

Reply:
If we introduce the option now, how do we handle the conflict when https://github.com/open-telemetry/opentelemetry-go/issues/4995 is resolved?

Reply:
I think it should first check if it is enabled via  `WithMinLevel` and if so then check if it enabled using API added via https://github.com/open-telemetry/opentelemetry-go/issues/4995 e.g. something like the `h.Logger.IsEnabled(...)`.

Reply:
Recap of sync conversation:

- Adding this will cause conflict when there are different people operating the code vs writing it
  - If a developer changes the logging level using `WithMinLevel` but the operator has set a level higher than that for the SDK that runs the application the developer will be confused why it is not seeing the logs they expect
  - If an operator changes the logging level of the SDK but an external library they depend on has used `WithMinLevel` to set the logging level the operator will not see any difference in logging level.
- Not adding this will mean that a user cannot increase/decreasing logging within their system package-by-package. They will only be able to set the level for the entire SDK.
  - This would be related to https://github.com/open-telemetry/opentelemetry-specification/pull/3877
  - A logging processor may be used in the SDK to address this if we build the SDK to support this situation.